### ISSUE TITLE
Security issues in s2ssh

### ISSUE THREAD

**Issue by jeandrek** [Scratch-Cloud-Services/scratch-cloud-services#3]
s2ssh is almost ready to release &mdash; I just need to test it first!
Can anybody see any security issues or vulnerabilities with s2ssh?
(feel free to tidy up the code if you like :stuck_out_tongue:)
Thanks!

https://github.com/Scratch-Cloud-Services/scratch-cloud-services/tree/s2ssh/s2ssh
https://github.com/Scratch-Cloud-Services/scratch-cloud-services/tree/s2ssh/s2ssh#security-concerns
https://github.com/Scratch-Cloud-Services/scratch-cloud-services/tree/s2ssh/s2ssh/s2ssh.js
- [x] make test account
- [ ] test project
- [ ] fix general bugs
- [ ] look for security vulnerabilities
- [ ] test project
- [ ] release version 1.0.0

**Comment by jeandrek** [comment]
@as-com @MegaApuTurkUltra @jayforsythe @hiccup01 @robinp7720 @BookOwl

**Comment by robinp7720** [comment]
I'll assume that the script just redirects stdout from openssh to scratch and scratch cloud vars to stdin? You are obviusly automaticly connecting to the host but I don't see anything to automaticly log you in. So, i'm guessing your gonna have to send your password in clear text through scratch? Assuming you don't use an ssh key agent

**Comment by MegaApuTurkUltra** [comment]
The whitelist is wayyyy too restrictive but I don't see any issues.

**Comment by jeandrek** [comment]
> I'll assume that the script just redirects stdout from openssh to scratch and scratch cloud vars to stdin? You are obviusly automaticly connecting to the host but I don't see anything to automaticly log you in. So, i'm guessing your gonna have to send your password in clear text through scratch? Assuming you don't use an ssh key agent

No, because a psuedo-terminal is allocated by OpenSSH to prompt for the password/key passphrase then it is done at the command line when running the script.

**Comment by jeandrek** [comment]
> The whitelist is wayyyy too restrictive but I don't see any issues.

Add stuff :package:
(have you used *nix?)

**Comment by robinp7720** [comment]
@Jonathan50 Ah ok. Seams reasonable

**Comment by jeandrek** [comment]
Ok
I've registered `@Jonathan50_test`
Time to test
<sup><sub>_after maths_</sub></sup>

**Comment by jeandrek** [comment]
Time to testtt............

**Comment by towerofnix** [comment]
Testtttting

**Comment by towerofnix** [comment]
```
192:~ towerofnix$ sudo npm install -g https://raw.githubusercontent.com/Scratch-Cloud-Services/scratch-cloud-services/s2ssh/s2ssh/s2ssh.tar.gz
npm http GET https://raw.githubusercontent.com/Scratch-Cloud-Services/scratch-cloud-services/s2ssh/s2ssh/s2ssh.tar.gz
npm http 200 https://raw.githubusercontent.com/Scratch-Cloud-Services/scratch-cloud-services/s2ssh/s2ssh/s2ssh.tar.gz
npm ERR! Error: ENOTDIR, open '/Users/towerofnix/tmp/npm-94172-Prn9yySe/1447717946476-0.37882198113948107/package/package.json'
npm ERR! If you need help, you may report this *entire* log,
npm ERR! including the npm and node versions, at:
npm ERR!     <http://github.com/npm/npm/issues>

npm ERR! System Darwin 15.0.0
npm ERR! command "node" "/usr/local/bin/npm" "install" "-g" "https://raw.githubusercontent.com/Scratch-Cloud-Services/scratch-cloud-services/s2ssh/s2ssh/s2ssh.tar.gz"
npm ERR! cwd /Users/towerofnix
npm ERR! node -v v0.10.26
npm ERR! npm -v 1.4.3
npm ERR! path /Users/towerofnix/tmp/npm-94172-Prn9yySe/1447717946476-0.37882198113948107/package/package.json
npm ERR! code ENOTDIR
npm ERR! errno 27
npm ERR! 
npm ERR! Additional logging details can be found in:
npm ERR!     /Users/towerofnix/npm-debug.log
npm ERR! not ok code 0
```

**Comment by towerofnix** [comment]
halp can i downlode it and run index.js

**Comment by jeandrek** [comment]
WOAH BE CAREFUL
I got that error too. If you are using Mac (which you are as it says `Darwin`) you should use that alternative command.
BE CAREFULLLLLLLLLL
1. unwhitelisted commands ran anyway (fixed by 9c2d3c03544f61e43be99a647758cd72b8838926, so get the **_latest_**)
2. can be crashed by entering Ctrl+D on Scratch. As this is not a recognised ASCII character a `-1` is outputted by the `get code` block. The `-` is not a digit, so it fails. (to be fixed)  
   (I find odd Ctrl+D adds a character. You also get a thing that says "currently cloud variables can only have numbers", due to the `-`)
3. Can anybody see any more??

Edit: oh cool referencing commits by sha1 works

**Comment by jeandrek** [comment]
Also bug: I can't exit `python`/`python3`/`perl`/`cat` :package:
The solution would be to replace `^D` with a EOF (U+0004, &#0004;)

<sup><sub>this is my first experience with computer security</sub></sup>

**Comment by jeandrek** [comment]
Wait I was adding replacing `^D` with EOF
vulnerability: people could exit the SSH session

**Comment by PolyEdge** [comment]
Why can't someone get one done in python? (not me I'm a noob at socket :unamused: )

**Comment by robinp7720** [comment]
@Dylan5797 I'm making a cloud api in python: https://github.com/robinp7720/ScratchCloud Its not even close to completed yet primarly because I can't find out how to get the session id.

**Comment by PolyEdge** [comment]
https://github.com/Dylan5797/ScratchAPI

**Comment by PolyEdge** [comment]
Looking for cloud token or session id?

**Comment by jeandrek** [comment]
> Looking for cloud token or session id?

https://scratch.mit.edu/projects/{id}/cloud-data.js

it's on that page

**Comment by PolyEdge** [comment]
Session ID maybe cookie "scratchsessionsid??? :stuck_out_tongue:

**Comment by PolyEdge** [comment]
I looked at source. I already know the cloud token trick.

**Comment by robinp7720** [comment]
Mine is able to change cloud vars if you use your existing session id.

**Comment by PolyEdge** [comment]
But where is it??!!??!!??

**Comment by PolyEdge** [comment]
I mean where did you get the token from

**Comment by PolyEdge** [comment]
Anybody know where you can find your (constant) cloud id?

**Comment by jeandrek** [comment]
> Anybody know where you can find your (constant) cloud id?

I already said:

> https://scratch.mit.edu/projects/{id}/cloud-data.js
> 
> it's on that page

just keep looking there, it's in a variable on that page
btw this is a bit offtopic

**Comment by PolyEdge** [comment]
I'm talking about cloud id not cloud token. So I put my "scratchsessionsid" into sessionid?
